src/defrag.c

```diff
@@ -935,6 +935,59 @@ void* defragStreamConsumerGroup(raxIterator *ri, void *privdata) {
     return cg;
 }
 
+/* Defrag a single idmpProducer's dict and linked list entries. */
+static void defragIdmpProducer(idmpProducer *producer) {
+    if (producer->idmp_dict == NULL) return;
+
+    dict *newdict = dictDefragTables(producer->idmp_dict);
+    if (newdict)
+        producer->idmp_dict = newdict;
+
+    idmpEntry **prevnext = &producer->idmp_head;
+    idmpEntry *entry = producer->idmp_head;
+
+    while (entry != NULL) {
+        idmpEntry *next = entry->next;
+        /* Delete before defrag to avoid use-after-free in hash */
+        dictDelete(producer->idmp_dict, entry);
+        idmpEntry *newentry = activeDefragAlloc(entry);
+        if (newentry) {
+            *prevnext = newentry;
+            if (producer->idmp_tail == entry)
+                producer->idmp_tail = newentry;
+            entry = newentry;
+        }
+        dictAdd(producer->idmp_dict, entry, NULL);
+        prevnext = &entry->next;
+        entry = next;
+    }
+}
+
+/* Defrag all IDMP producers and their dict/linked list entries. */
+void defragIdmpProducers(stream *s) {
```

Comment:
what about defragStreamIdmpProducers, at first glance, i confused defragIdmpProducer() with it.